### DISCUSSION TITLE
What is happening here?

### DISCUSSION THREAD

**Discussion by 9mm** [neovide/neovide#2074] [Q&A] [answered]
This is a bit hard to describe as I can't pinpoint when this happens. It's about 2-3 times a day, something I'm doing but I never notice what I did previously

Normally I just do `:w :w :w` all day long to save, and it works...

But then, sometimes....

I will do `:w` and then it will require me to hit `ENTER` on some confirmation. I'm trying to get it to occur again so I can see the exact message, but if I recall its nothing helpful, its just showing the name of the file.

So normally it might say:

```
init.lua .... written
```

but then sometimes it will say

```
init.lua ... PRESS ENTER
```

Or something to that effect.

If I close the tab/buffer, it does NOT fix the issue for that file only. oddly enough, other files will work fine.

I have to completely quit neovide, and then it will go away on that one file, which leads me away from the idea of some kind of swap file weirdness like I thought originally.

Next time it happens I will get an exact screenshot

**Comment by 9mm** [comment]
Ok i got it to trigger

<img width="855" alt="image" src="https://github.com/neovide/neovide/assets/1386207/17e7f1a9-72c6-46a6-b45e-39ee0491149e">

It seems like it might be after renaming a buffer with neotree. Regardless of whether this is a neotree bug or not (which it probably is) im still wondering what this actually is, or if there is a way i can just disable it in neovim without having to go down rabbit hole in neotree

**Comment by 9mm** [comment]
its almost certainly related to neovide messing with buffers after looking further, in a way where i really just need to fix it there. sigh

**Reply by fredizzimo** [reply]
Does the following mapping do the same thing? (In an new Neovide instance, since it messes up the default mappings)

`map a <cmd>echomsg "hello" <bar> echomsg "world"<cr>`

It seems like what your are seeing is just the standard behaviour when there are multiple messages shown. Although it should disappear when you press enter. And at least this variation does for me. I also tested with `cmdheight=0` and that works correctly for me too.

I'm also confused when you talk about file specific messages, since the messages should be global no matter which file you have active, unless it's some plugin that is displaying them.

**Reply by 9mm** [reply]
@fredizzimo yep that causes the same issue:

<img width="429" alt="image" src="https://github.com/neovide/neovide/assets/1386207/6182f993-a8fc-4630-a99e-8e701839dc88">

And sorry, it wasnt specific messages, moreso its one message but on SOME files when I do `:w` it shows the 'press enter' message, while on a different file in a different tab/buffer a `:w` works like normal

I will dig into it some more with  neotree

**Reply by fredizzimo** [reply]
If that causes the issue, then it's not Neotree. 

For me it shows the message when I press any key it closes, which is normal. So maybe it's some other plugin that messes it up. Can you try with `neovide -- --clean`?

**Reply by 9mm** [reply]
Hmm wait I think I'm misunderstanding what the expected and unexpected behavior should be for your mapping. I dont want to give a false positive.

When I paste `map a <cmd>echomsg "hello" <bar> echomsg "world"<cr>` into the command line and hit enter, it makes it where any time i press the `a` key in normal mode it shows 2 lines (hello line 1, world line 2) and the 3rd line is "press enter or ...."

I can hit enter/any key and the message goes away

the original problem is that usually when you write `:w` and hit enter a single time it just writes the file and thats it. But when its bugged out, it requires a second `<CR>` because the message shows up, so its more like `:w<cr><cr>`

**Reply by 9mm** [reply]
Ok Im trying to poke around now. ive disabled every single plugin. One very bizarre thing i noticed (im not 100% this is true but 80% true as of all my experiemnts), it only happens with files that are at least 2 levels deep in folders, instead of 1 folder.

Also, i uncommented the following lines after commenting out every single thing, and i got this

```
vim.opt.updatetime = 50
vim.opt.timeout = true
vim.opt.timeoutlen = 600
vim.opt.shortmess = 'I'
```

<img width="901" alt="image" src="https://github.com/neovide/neovide/assets/1386207/abfd1e20-c617-402e-a376-51bffbb63275">

I have also got it where it triggers on opening a file and not just saving the file.

**Reply by fredizzimo** [reply]
Then I was misunderstanding your original issue. I thought it was not possible to dismiss the message at all.

But yes, that is standard Neovim behaviour any time more than two messages are written, a prompt is shown. 

There are some `shortmess` options, but I don't think any of them will fix it https://neovim.io/doc/user/options.html#'shortmess'. 

You can also set the `cmdheight` to something higher as recommended here https://neovim.io/doc/user/message.html#hit-enter-prompt.

There are also these two issues
* https://github.com/neovim/neovim/issues/1029
* https://github.com/neovim/neovim/issues/22478

**Comment by 9mm** [comment]
@fredizzimo ok ... i actually was going to make an issue on neotree, but then of course remembered your advice and it actually does NOT occur in `nvim` no matter how hard i try, it only occurs in neovide. Before making an issue I wanted to put this here though to see if you think it deserves making a separate issue. Im going to try it about 100 more times to be sure, but here is a loom I made showing what happens in neovide. im using the improve render loop branch.

When you watch this, keep in mind there are 3 separate unexpected bad behaviors. The first behavior, #1 below, is the most common one. But i've seen the other 2 occur at least 2 or 3 times.

It almost seems like some kind of buffering issue or something as if it gets stuck "off by one" and then is always off by one until I close

1. (most common) - Typing `:w` in a file requires hitting `<CR>` an extra time because there is an unexpected "press enter" message that shows up
2. (somewhat rare) - hitting `<CR>` on neotree to open a file requires hitting `<CR>` an extra time. I actually caught this one on the loom
3. (somewhat rare) - I noticed TWO status bar lines after saving a single file with `:w` is if its buffering multiple saves or something. This is the image I put above. I didnt catch this on the loom, but i did take a screenshot of it above. Note that I didnt use `:wa` for multiple files, just simply `:w`

![image](https://github.com/neovide/neovide/assets/1386207/4eebe0d3-fe8a-4cc2-9dd3-1b95243461ff)

Here is loom

https://www.loom.com/share/faf95ea247bd43b2811275e164dd7d30

**Comment by 9mm** [comment]
I also noticed this is backgraces log in the same folder. I have no idea if this is related. It might be because I believe the timestamp matches about 20 mins ago. Theres only 2 traces in it though, and ive been reproducing it many times... so .. maybe its not related

```
@ -0,0 +1,28 @@
2023-10-11 18:40:19 - Neovide panicked with the message 'called `Result::unwrap()` on an `Err` value: SendError { .. }'. (File: src/event_aggregator.rs; Line: 51, Column: 28)
   0: __mh_execute_header
   1: __mh_execute_header
   2: __mh_execute_header
   3: __mh_execute_header
   4: __mh_execute_header
   5: __mh_execute_header
   6: __mh_execute_header
   7: __mh_execute_header
   8: __mh_execute_header
   9: __mh_execute_header
  10: __mh_execute_header
  11: __pthread_joiner_wake

2023-10-11 18:43:32 - Neovide panicked with the message 'called `Result::unwrap()` on an `Err` value: SendError { .. }'. (File: src/event_aggregator.rs; Line: 51, Column: 28)
   0: __mh_execute_header
   1: __mh_execute_header
   2: __mh_execute_header
   3: __mh_execute_header
   4: __mh_execute_header
   5: __mh_execute_header
   6: __mh_execute_header
   7: __mh_execute_header
   8: __mh_execute_header
   9: __mh_execute_header
  10: __mh_execute_header
  11: __pthread_joiner_wake
```

**Reply by fredizzimo** [reply]
You are not the only one that has reported this. There's a problem with some of my new changes, that needs fixing.

**Comment by 9mm** [comment]
Interestingly, when I rename something outside of neovide, I dont get any confirmation at all

<img width="1130" alt="image" src="https://github.com/neovide/neovide/assets/1386207/cfe98917-1696-47dd-9dd1-4f8efaa8bd86">

Im assuming this is the desired behavior as ive always thought hitting enter there is super annoying

In neovide I get this 

<img width="923" alt="image" src="https://github.com/neovide/neovide/assets/1386207/96abc865-059b-43dc-907b-2bf9238ac033">

Its almost like by getting that "show enter" to appear, everything after that becomes kind of funky

**Comment by 9mm** [comment]
I mean, this issue makes NO sense.

If I open Neovide, and just ignoring this issue entirely and only focusing on the question "If I rename a file to itself, which generates an 'already exists' error, Do i get the "press enter" confirmation error?

If its a file ONE level deep, i do not get the confirmation.

If its multiple layers deep I get it 90% of the time (but not 100%). Id make another loom but thats pretty straightforward to understand.

**Comment by fredizzimo** [comment]
I'm not sure, but the press enter message is sent if there are two messages in the same frame of drawing, if they are sent on different frames it's not shown. The terminal mode neovim probably redraws a little bit differently, which explains the differences, and also why it's a bit inconsistent.

The actual press enter message and the reading of the kestroke all happens inside Neovim, so there's not much we can do about that.

The only way of getting rid of the message, that I can think of is to do something like `:silent w` (the shortest is `:sil w`). Or a silent mapping. Or something crazy like `cabbrev w silent w` but that could have other side effects.

**Reply by 9mm** [reply]
Is there  a way I can see what the other message is? In the instance of a "rename" i suspect the only message that would show is the "already exists" message. I disabled all other plugins and config, so Im wondering if i could somehow log the other message to see what else could be causing it

**Reply by fredizzimo** [reply]
Yes, they should show up in `:messages`

**Reply by 9mm** [reply]
There is only a single message in there.

Here is opening fresh vim, renaming a file to itself to generate a confirmation, and immediately typing `:messages` after... theres 1 message. Heres the output of `:messages`:

<img width="879" alt="image" src="https://github.com/neovide/neovide/assets/1386207/7e587f24-7dd3-485f-b367-1f0e6d763c6a">

Heres what happens if I pick a file 1 level deep so theres NO confirmation. It appears the output is the same

<img width="812" alt="image" src="https://github.com/neovide/neovide/assets/1386207/3e8025cf-a24f-4722-a842-c809d21950e6">

Here is NVIM, no confirmation:

<img width="1079" alt="image" src="https://github.com/neovide/neovide/assets/1386207/ff9f46c9-b4e6-4ab0-95f3-84da0c7468c6">

**Reply by fredizzimo** [reply, marked as answer]
I see what the problem is now, it’s wrapping to the next line, so two lines has to be output. You can try some shortmess options to make it shorter

**Reply by fredizzimo** [reply]
Ah, no it’s from Neotree, so they need to make it shorter, or disable it

**Reply by 9mm** [reply]
AHHHHH!!!!!!!!!!!!!!!!!!!!!! Dude YES. Good call. Now it makes sense why it would onyl work for some files multiple levels deep but not all of them. OMG!

Thank you!!!!! I'll have a look at some shortmess stuff

**Reply by 9mm** [reply]
Youre right, if i use neovim (not neovide) and make super small, it causes a confirmation.

I have no idea if thats related to the original issue but im pretty sure it is. It probably had nothing to do with renaming it was because i kept adding `2` the end of the file making it much longer

i can open nvim now and simply type `:w` and if the window is small it causes a confirmation

**Reply by fredizzimo** [reply]
I’m glad that we at least found out what the issue is. But there are very few options for fixing it. Noice is one https://github.com/folke/noice.nvim. We could also implement something like that built-in, but that would automatically disable noice, so some kind of option is probably needed. And of course it’s a quite big thing to implement.
